### PR TITLE
`GRPOTrainer/async`: fix prefix EOS slicing for tool suffix (with Qwen3/3.5 type of chat templates)

### LINKED ISSUE #5317: Missing `\n` token between completion and tool suffix in `GRPOTrainer._get_tool_suffix_ids` for Qwen3 and Qwen3.5
### Reproduction

In Qwen3 or Qwen3.5 chat template, we'll have `\n` token between EOS token of the completion `<|im_end|>` and the start of the tool messages `<|im_start|>user\n<tool_response>`.

As generation always ends at EOS token, `completion_ids` will end with the `<|im_end|>` and with the current implementation of `GRPOTrainer._get_tool_suffix_ids`, `suffix_ids` will start with `<|im_start|>user\n<tool_response>`.

Therefore, we'll miss the `\n` token in `prompt_completion_tool_ids = prompt_ids + completion_ids + suffix_ids`.

```
What the chat template expects:
  ...<|im_end|> \n <|im_start|>user\n<tool_response>...
       ↑         ↑        ↑
  completion   THIS    suffix_ids
    ends      TOKEN     starts
              IS LOST

Where each part comes from:
  completion_ids:  ...<|im_end|>          ← generation stops at EOS, no \n
  prefix_ids:      ...<|im_end|>\n        ← includes the \n
  suffix_ids:      <|im_start|>user\n...  ← starts after prefix, \n already consumed

Result of prompt_ids + completion_ids + suffix_ids:
  ...<|im_end|><|im_start|>user\n...     ← \n is missing!
```

```python
>>> from transformers import AutoTokenizer

>>> from trl.chat_template_utils import get_training_chat_template

>>> tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-0.6B")

>>> training_chat_template = get_training_chat_template(tokenizer)

>>> tool_messages = [
...     {"role": "tool", "tool_call_id": "call_001", "content": '{"temperature": 18, "condition": "partly cloudy"}'}
... ]


>>> dummy_messages = [{"role": "user", "content": "dummy"}, {"role": "assistant", "content": "dummy"}]

>>> prefix_ids = tokenizer.apply_chat_template(
...     dummy_messages,
...     add_generation_prompt=False,
...     chat_template=training_chat_template,
...     return_dict=True,
... )["input_ids"]

>>> tokenizer.decode(prefix_ids)
'<|im_start|>user\ndummy<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\ndummy<|im_end|>\n'

>>> full_ids = tokenizer.apply_chat_template(
...     dummy_messages + tool_messages,
...     add_generation_prompt=True,
...     chat_template=training_chat_template,
...     return_dict=True,
... )["input_ids"]

>>> tokenizer.decode(full_ids)
'<|im_start|>user\ndummy<|im_end|>\n<|im_start|>assistant\n<think>\n\n</think>\n\ndummy<|im_end|>\n<|im_start|>user\n<tool_response>\n{"temperature": 18, "condition": "partly cloudy"}\n</tool_response><|im_end|>\n<|im_start|>assistant\n'

>> suffix_ids = full_ids[len(prefix_ids) :]

>>> tokenizer.decode(suffix_ids)
'<|im_start|>user\n<tool_response>\n{"temperature": 18, "condition": "partly cloudy"}\n</tool_response><|im_end|>\n<|im_start|>assistant\n'
```

I think we can have 2 options to avoid this:
1. Either right strip `prefix_ids` until reach EOS token
2. Or update the training chat_templates in `get_training_chat_template` to get rid of the `\n` token at the end when `add_generation_prompt=False`

### System Info

trl: `pip install git+https://github.com/huggingface/trl.git@ebdfe826f6b1fcc2e82a5e01b9875f9688a56c52`

```
- Platform: Linux-6.8.0-83-generic-x86_64-with-glibc2.35
- Python version: 3.11.15
- TRL version: 1.0.0.dev0
- PyTorch version: 2.10.0
- accelerator(s): NVIDIA H200
- Transformers version: 5.3.0
- Accelerate version: 1.13.0
- Accelerate config: not found
- Datasets version: 4.5.0
- HF Hub version: 1.7.1
- bitsandbytes version: 0.49.1
- DeepSpeed version: 0.18.2
- Liger-Kernel version: 0.7.0
- LLM-Blender version: not installed
- OpenAI version: 2.24.0
- PEFT version: 0.18.1
- vLLM version: 0.17.0
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by casinca**
# What does this PR do?

Hi,
this should be a Fix #5317 

I went with option 1 (model agnostic imho) mentioned by @mattbui in the issue, and mirrored it to the new exp async GRPO. 
Generated a test case, in case.

## Before submitting
- [] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes how tool-result token suffixes are computed in both `GRPOTrainer` and experimental async GRPO; mistakes here could break tool-calling flows or cause subtle tokenization mismatches for certain chat templates.
> 
> **Overview**
> Fixes tool-call prompt reconstruction for chat templates (notably Qwen3/Qwen3.5) that emit an EOS token followed by extra trailing tokens (e.g. newline). When deriving tool-result `suffix_ids`, the code now trims the prefix to the last EOS before slicing and validates the EOS-trimmed prefix matches the full tokenization, preventing incorrect suffix extraction.
> 
> Adds a regression test (`test_get_tool_suffix_ids_eos_newline_boundary`) covering Qwen3 and Qwen3.5 tokenizers (with version guard) to ensure `_get_tool_suffix_ids` returns the expected suffix under this EOS+newline boundary condition.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a8e0f3ce8bf99ca63d91b7ed963526ead7c7b6fb. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1394,9 +1394,20 @@ def _get_tool_suffix_ids(self, tool_messages):
             return_dict=False,
             **self.chat_template_kwargs,
         )
-        if not full_ids[: len(prefix_ids)] == prefix_ids:
-            raise ValueError("Unexpected tokenization: the prefix IDs are not a prefix of the full IDs.")
-        return full_ids[len(prefix_ids) :]
+
+        # Some chat templates (notably Qwen3/Qwen3.5) render "...<|im_end|>\\n" after an assistant/tool block.
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
        # Some chat templates (notably Qwen3/Qwen3.5) render "...<|im_end|>\n" after an assistant/tool block.
```

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1394,9 +1394,20 @@ def _get_tool_suffix_ids(self, tool_messages):
...
+        # When we compute `suffix_ids` by slicing `full_ids`, we must align the slicing boundary to
+        # EOS (not EOS + newline).
+        prefix_ids_for_suffix = prefix_ids
+        if self.eos_token_id is not None and self.eos_token_id in prefix_ids:
+            last_eos_idx = max(i for i, tok_id in enumerate(prefix_ids) if tok_id == self.eos_token_id)
+            if last_eos_idx < len(prefix_ids) - 1:
+                prefix_ids_for_suffix = prefix_ids[: last_eos_idx + 1]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
        last_eos_idx = max(i for i, tok_id in enumerate(prefix_ids) if tok_id == self.eos_token_id)
        prefix_ids_for_suffix = prefix_ids[:last_eos_idx + 1]
```

The first condition can't be False. Or at least shouldn't be.

The second one is not necessary

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -1394,9 +1394,20 @@ def _get_tool_suffix_ids(self, tool_messages):
...
+
+        if full_ids[: len(prefix_ids_for_suffix)] != prefix_ids_for_suffix:
+            raise ValueError("Unexpected tokenization: the EOS-trimmed prefix IDs are not a prefix of the full IDs.")
+
+        return full_ids[len(prefix_ids_for_suffix) :]
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
        last_eos_idx = max(i for i, tok_id in enumerate(prefix_ids) if tok_id == self.eos_token_id)
        prefix_ids = prefix_ids[: last_eos_idx + 1]

        if full_ids[: len(prefix_ids)] != prefix_ids:
            raise ValueError("Unexpected tokenization: the EOS-trimmed prefix IDs are not a prefix of the full IDs.")

        return full_ids[len(prefix_ids) :]
```

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: simpler

File: tests/test_grpo_trainer.py [hunk 4]
```diff
@@ -2641,6 +2642,67 @@ def test_single_reward_model_with_single_processing_class(self):
         assert len(trainer.reward_processing_classes) == 1
         assert trainer.reward_processing_classes[0] == single_processing_class
 
+    @pytest.mark.parametrize(
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]: We should either make the method static and test it, or remove this test (having such mocking is a nightmare to maintain). I recommend that we simply remove this test for now

Comment (casinca) [reply tests/test_grpo_trainer.py]: Ah ok, np i removed it

File: trl/experimental/async_grpo/async_rollout_worker.py [hunk 5]
```diff
@@ -573,10 +573,17 @@ def _build_messages_suffix_ids(self, messages: list[dict[str, Any]]) -> list[int
...
-            raise ValueError("Failed to construct message suffix in token space.")
-        return prefix_and_messages_ids[prefix_len:]
+
+        # Some chat templates (notably Qwen3/Qwen3.5) render "...<|im_end|>\n" after an assistant/tool block.
+        # When we compute `suffix_ids` by slicing `full_ids`, we must align the slicing boundary to
+        # EOS (not EOS + newline).
+        last_eos_idx = max(i for i, tok_id in enumerate(prefix_ids) if tok_id == self.tokenizer.eos_token_id)
+        prefix_ids = prefix_ids[: last_eos_idx + 1]
+
+        if prefix_and_messages_ids[: len(prefix_ids)] != prefix_ids:
```
Comment (qgallouedec) [inline trl/experimental/async_grpo/async_rollout_worker.py:583]: strange that we don't use the same variable names than in GRPO, we should align this in the future

**Review by qgallouedec — APPROVED** [review]
LGTM! thanks a lot for the fast fix!

File: trl/experimental/async_grpo/async_rollout_worker.py (hunk 5 above)
Comment (casinca) [reply trl/experimental/async_grpo/async_rollout_worker.py:583]: Followed in https://github.com/huggingface/trl/pull/5332
